### PR TITLE
Add half-cones and Ipelet

### PR DESCRIPTION
**Pull request by freud14**
Hi, I made a small Ipelet using the Cone-Based Spanners package. I also included a way to construct half-theta and half-yao graphs. Thank you.

Small-feature page [here](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Cone_spanner_half_cones).

Note: pre-approved 20 dec. 2016

### TODO
- [x] update `changes.html`

### CODE REVIEW

**Comment by ws4u** [comment]
Thank @freud14 for the contribution. I looked at the code for half yao/theta graph construction. My comments are as follows:

- The API is properly modified by adding a default parameter to indicate full or half construction. I think this API modification is necessary, and kept at minimum.
- The code for the actual construction is concise. I didn't download and run the code, but according to my reading of the code, I think the code can correctly construct half yao/theta graphs. 
- If the code is to be merged into master, the user manual for the cone-based spanners should be updated to document the updated API.
- There is a small typo in the code comments: wheter -> whether

For the added Ipelet, I'm not good at it. However, @sloriot may have time to comment on it.

**Comment by freud14** [comment]
Thank you @ws4u for your response. I made a commit to fix the small typo. For the user manual, what do you suggest? IMO, a small mention in the Definitions Section would do. Thank you.

**Comment by freud14** [comment]
Hi,

I notice that when two points are on the horizontal line, because of the imprecision of the Exact_predicates_inexact_constructions_kernel, I do not get what I am supposed to. I tested with the Exact_predicates_exact_constructions_kernel_with_root_of and it seems to work fine. I read [there](http://doc.cgal.org/latest/CGAL_ipelets/index.html#Chapter_CGAL_Ipelets) that it is fine to use this kernel in an Ipelet, but I get a linking error in Ipe when I try to use the Ipelet unless I do `export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/usr/local/lib/` before opening Ipe. Any idea?

Furthermore, I just made some tests and it seems that the documentation is inaccurate. It says that a point on l_i is part of the cone c(u, i). However, according to my tests in Ipe, it's the points on l_{i+1} that seem to belong to c(u, i).

Thank you.

Frédérik

**Comment by ws4u** [comment]
@freud14 For your second question, I'll experiment myself and examine the code again. This week I'm a little busy, but I'll get back to you next week. Thanks for your feedback.

**Comment by lrineau** [comment]
@sloriot As the unofficial maintainer of our ipelets, can you please check this PR?

**Comment by lrineau** [comment]
Its seems the API and documentation of `Cone_spanners_2` is amended. In my opinion there should be a corresponding small feature page.

**Comment by ws4u** [comment]
@freud14 I tested and examined my code again. You are correct. In the current implementation, l_{i+1}, not l_i, belongs to c(u, i). To get an implementation that l_i belongs to c(u,i), the following two lines of code in "Less_by_direction_2.h" 
`  // define a rotation of counter clockwise 90`
`  Transformation ccw90(0, -1, 1,  0);`
should be changed to:
`  // define a rotation of clockwise 90`
`  Transformation cw90(0, 1, -1,  0);`
This will perfectly fix the problem. I've tested about it. If you have further questions, please feel free to let me know.

**Comment by freud14** [comment]
@ws4u Thank you for your answer. The question is: do we want to change the implementation or the documentation?

**Comment by ws4u** [comment]
@freud14 I will create a pull request to change the implementation, and also add code comments to the implementation on what to do if users want l_{i+1} on c(u,i).

**Comment by sloriot** [comment]
The ipelet compiles and run. The code looks good too.

**Comment by freud14** [comment]
@sloriot I have a small problem that I mentionned in this comment: https://github.com/CGAL/cgal/pull/1673#issuecomment-260390903 Do you have an answer for it? Thank you.

**Comment by sloriot** [comment]
@freud14  I guess you have a link issue since Core is needed when `Exact_predicates_exact_constructions_kernel_with_root_of` is used.

**Comment by freud14** [comment]
@sloriot Hmm that's weird. It wasn't working correctly, but now it works. Thank you. I just committed the change of kernel if you want to verify it.

**Comment by sloriot** [comment]
Runs correctly here too.

**Comment by freud14** [comment]
> In my opinion there should be a corresponding small feature page.

@lrineau Do you want me to do that? Do you have instructions for that?

**Comment by sloriot** [comment]
I suggest that you do the modification asked by @ws4u in the user manual (updating the definition section and mention half-XXX in the description in `PackagesDescription.txt`). When @ws4u is happy with the modifications, we'll deal with the small feature as you don't have access to our internal resources (this is a human readable page summarizing API change and doc changes).
I have a naive question (that also applies to Yao-graph) after reading the Wikipedia page of Theta-graph. I guess you compute one of _even_ and _odd_ Theta-graph (_even_ IIUC). Why not offering the possibility to also compute the complement?

**Comment by freud14** [comment]
> Why not offering the possibility to also compute the complement?

That's a good idea. So, instead of a boolean, it could be an enum with items {EVEN_CONES, ODD_CONES, ALL_CONES} ?

**Comment by sloriot** [comment]
Looks good to me. @ws4u what do you think?

**Comment by ws4u** [comment]
@sloriot Looks good to me as well. For the Ipelet part, where should we document it, in both CGAL Ipelets chapter and Cone-based Spanners chapter?

@freud14 In modifying `PackagesDescription.txt`, feel free to ask me if you have any questions. Please add you as an author in the user manual.

**Comment by sloriot** [comment]
The ipelet should be mentioned in the user manual of the ipelets.

**Comment by freud14** [comment]
So, I added the choice for even or odd cones and I added all the necessary documentation. However, I don't know why but the elements of the enum don't seem to display in the doc. I don't know what I did wrong. I tried to copy the enum in the CGAL/enum.h files of Kernel_23 and it was working fine so I'm wondering if there is some configuration to add to the Cone_spanners_2 package.

**Comment by ws4u** [comment]
Thanks @freud14. I suggest the following changes to your submission:

- Rename the file "Cone_spanners_2_enum.h" to "Cone_spanners_enum_2.h"
- Rename the enum type `Half` to `Cones_selected`
- Rename the member variable `half` to `cones_selected`
- Change the comment `/* Store whether this is an half-theta graph */` to `/* Store whether even, odd or all cones are selected to construct graph */`
- The user manual is converted from the "Cone_spanners_2.txt" file. I'm sorry for mentioning it wrongly as "PackageDescription.txt"  in my last comment. Pls add a new subsection into this file after the subsection titled "Constructing Graphs Exactly or Inexactly and Generating Gnuplot Files". This new subsection should be titled sth like "Constructing Half Theta or Yao Graphs". In this new subsection, pls briefly describe the  following:
-- By default, a full theta or yao graph is constructed.
-- To construct a half theta or yao graph, users should include "Cone_spanners_enum_2.h" in the cpp file, and supply one more parameter when initializing the functor.
-- Adding a code snippet for this will be OK. No need to include an entire file.
- Add authorship to the files where you make contributions.

For the above suggestions, you don't need to follow them strictly, but can have your own creative ideas. For the above renamings, pls wait for @sloriot 's comment before carrying them out.

**Comment by ws4u** [comment]
@freud14, @sloriot  To make the elements of the enum display in the doc, a compilation on the doc is needed. I'm not sure whether this is available to public contributors.

**Comment by freud14** [comment]
@ws4u Just so you know, I already added a small paragraph in the Definitions Section. I will add a section like you suggest.

**Comment by freud14** [comment]
I'm already able to compile the doc. When I click on the type of my parameter, it just stays on this [page](http://imgur.com/zH3t6mL.png) and doesn't display the elements of the enum.

**Comment by ws4u** [comment]
Can you try adding comments for each element?

File: CGAL_ipelets/demo/CGAL_ipelets/cone_spanners.cpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+  "Draws an half-Yao-graph with the odd of k cones.",
+  "Draws k cones around the points.",
+};
+
+class Cone_spanners_ipelet
+  : public CGAL::Ipelet_base<Kernel,6> {
+public:
+  Cone_spanners_ipelet()
+    :CGAL::Ipelet_base<Kernel,6>("Cone Spanners",labels,hmsg){}
+  void protected_run(int);
```
Comment (sloriot) [inline CGAL_ipelets/demo/CGAL_ipelets/cone_spanners.cpp]: 6 -> 7 I guess

File: Cone_spanners_2/include/CGAL/Cone_spanners_2_enum.h [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+#ifndef CONE_SPANNERS_2_ENUM_H
+#define CONE_SPANNERS_2_ENUM_H
+
+namespace CGAL {
+  /*! \ingroup PkgConeBasedSpanners
+
+   \brief An enum of the types of cone spanners.
+   */
+  enum Half { EVEN_CONES = 0, ODD_CONES = 1, ALL_CONES = 2 };
```
Comment (sloriot) [inline Cone_spanners_2/include/CGAL/Cone_spanners_2_enum.h]: The name of the enum is too generic.

File: Cone_spanners_2/include/CGAL/Construct_theta_graph_2.h [hunk 3]
```diff
@@ -90,10 +94,13 @@ class Construct_theta_graph_2 {
      \param initial_direction  A direction denoting one of the rays dividing the
                    cones. This allows arbitary rotations of the rays that divide
                    the plane.  (default: positive x-axis)
+     \param half_theta  Indicates whether all the cones are used or just half of
+                        them (default: all cones).
```
Comment (sloriot) [inline Cone_spanners_2/include/CGAL/Construct_theta_graph_2.h]: no need to mention the default

File: Cone_spanners_2/include/CGAL/Construct_yao_graph_2.h [hunk 4]
```diff
@@ -85,10 +89,13 @@ class Construct_yao_graph_2 {
       \param initial_direction  A direction denoting one of the rays dividing the
                    cones. This allows arbitary rotations of the rays that divide
                    the plane.  (default: positive x-axis)
+     \param half_yao  Indicates whether all the cones are used or just half of
+                      them (default: all cones).
```
Comment (sloriot) [inline Cone_spanners_2/include/CGAL/Construct_yao_graph_2.h]: No need to mention the default

File: Cone_spanners_2/include/CGAL/Cone_spanners_2_enum.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+// This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
+// WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+//
+// $URL$
+// $Id$
+//
+//
+// Authors: Weisheng Si, Quincy Tse
+
+/*! \file Cone_spanners_2_enum.h
```
Comment (sloriot) [inline Cone_spanners_2/include/CGAL/Cone_spanners_2_enum.h]: Rename the file `Cone_spanners_enum_2.h`

**Comment by freud14** [comment]
I did the modifications asked. I added comments on each element of the enum so that they can display in the doc. However, when trying to reference the enum with quotes (\`), the link is not made in the doc. I don't know why.

**Comment by freud14** [comment]
Done.

**Comment by ws4u** [comment]
Great, I'm ok with the doc now. For the back quote (`), it doesn't work occasionally for me as well. I don't know why.

**Comment by lrineau** [comment]
@sloriot might know how to fix the auto-linking in the documentation.

**Comment by freud14** [comment]
Hi, is there anything remaining I have to do? Thank you.

**Comment by sloriot** [comment]
Sorry I forgot to submit the small feature. I'll do it right away.

File: Cone_spanners_2/include/CGAL/Construct_yao_graph_2.h [hunk 6]
```diff
@@ -16,7 +16,7 @@
 // $Id$
 //
 //
-// Authors: Weisheng Si, Quincy Tse
+// Authors: Weisheng Si, Quincy Tse, Frédérk Paradis
```
Comment (sloriot) [inline Cone_spanners_2/include/CGAL/Construct_yao_graph_2.h]: typo here

**Comment by sloriot** [comment]
TODO
- [ ] update `changes.html`

**Comment by sloriot** [comment]
@ws4u you can check out and pre-approved the [small feature page](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Cone_spanner_half_cones).

File: Cone_spanners_2/include/CGAL/Construct_yao_graph_2.h (hunk 6 above)
Comment (freud14) [inline Cone_spanners_2/include/CGAL/Construct_yao_graph_2.h]: Fixed.

**Comment by ws4u** [comment]
@sloriot I have no problem with the small feature page.

**Comment by lrineau** [comment]
There are `-Wreorder` warnings with g++:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.10-Ic-81.shtml#Cone_spanners_2
```
.../include/CGAL/Construct_theta_graph_2.h:103:48: warning: field 'rays' will be initialized after field 'cones_choice' [-Wreorder]
                            ): cone_number(k), rays(std::vector<Direction_2>(k)), cones_choice(cones_selected)
                                               ^
```

**Comment by freud14** [comment]
Fixed. Btw, do you receive a notification when I push a commit?

**Comment by lrineau** [comment]
Thanks. Yes, I watch the whole CGAL/cgal repo, and that means all pull-requests. You can also see me in the list of participants of this discussion.

This PR will be re-tested tonight.

**Comment by sloriot** [comment]
@freud14 I can make the change to `changes.html` if you add me to your collaborators for this repo.

**Comment by freud14** [comment]
@sloriot You're added. The branch is cone-spanners-ipelet.

**Pull request merged**